Source/Diagnostics/BoostedFrameDiagnostic.cpp

```diff
@@ -216,10 +686,55 @@ writeLabFrameData(const MultiFab* cell_centered_data,
     VisMF::SetHeaderVersion(current_version);    
 }
 
+#ifdef WARPX_USE_HDF5
 void
 BoostedFrameDiagnostic::
-writeParticleData(const WarpXParticleContainer::DiagnosticParticleData& pdata, const std::string& name,
-                  const int i_lab)
+writeParticleDataHDF5(const WarpXParticleContainer::DiagnosticParticleData& pdata,
+                      const std::string& name, const std::string& species_name)
+{
+    auto np = pdata.GetRealData(DiagIdx::w).size();
+    
+    Vector<long> particle_counts(ParallelDescriptor::NProcs(), 0);
+    Vector<long> particle_offsets(ParallelDescriptor::NProcs(), 0);
+    
+    ParallelAllGather::AllGather(np, particle_counts.data(), ParallelContext::CommunicatorAll());
+    
+    long total_np = 0;
+    for (int i = 0; i < ParallelDescriptor::NProcs(); ++i) {
+        particle_offsets[i] = total_np;
+        total_np += particle_counts[i];
+    }
+
+    if (total_np == 0) return;
+    
+    long old_np = 0;
+    if (ParallelDescriptor::IOProcessor())
+    {
+        for (int k = 0; k < static_cast<int>(particle_field_names.size()); ++k)
+        {
+            std::string field_path = species_name + "/" + particle_field_names[k];
+            old_np = output_resize_particle_field(name, field_path, total_np);
+        }
+    }
+
+    ParallelDescriptor::ReduceLongMax(old_np);
```

Comment:
It looks like this function call also plays the role of an MPI barrier (between the `resize`, which done only by the IO proc, and the `write` which is done by all procs. Is that correct?
If yes, it might be worth stating this explicitly in a comment, e.g. above this line.